pkg/rtc/participant.go

```diff
@@ -272,84 +276,91 @@ func (p *ParticipantImpl) ConnectedAt() time.Time {
 
 // SetMetadata attaches metadata to the participant
 func (p *ParticipantImpl) SetMetadata(metadata string) {
-	p.lock.Lock()
-	changed := p.params.Grants.Metadata != metadata
-	p.params.Grants.Metadata = metadata
-	p.lock.Unlock()
+	grants := p.ClaimGrants()
+	changed := grants.Metadata != metadata
+	grants.Metadata = metadata
+	p.grants.Store(grants)
 
 	if !changed {
 		return
 	}
 
-	if p.onMetadataUpdate != nil {
-		p.onMetadataUpdate(p)
+	if p.onParticipantUpdate != nil {
+		p.onParticipantUpdate(p)
 	}
 	if p.onClaimsChanged != nil {
 		p.onClaimsChanged(p)
 	}
 }
 
 func (p *ParticipantImpl) ClaimGrants() *auth.ClaimGrants {
-	p.lock.RLock()
-	defer p.lock.RUnlock()
-	return p.params.Grants
-}
-
-func (p *ParticipantImpl) SetPermission(permission *livekit.ParticipantPermission) {
-	p.lock.Lock()
-	p.permission = permission
+	return p.grants.Load().(*auth.ClaimGrants)
+}
+
+func (p *ParticipantImpl) SetPermission(permission *livekit.ParticipantPermission) bool {
+	if permission == nil {
+		return false
+	}
+	grants := p.ClaimGrants()
+	video := grants.Video
+	hasChanged := video.GetCanSubscribe() != permission.CanSubscribe ||
+		video.GetCanPublish() != permission.CanPublish ||
+		video.GetCanPublishData() != permission.CanPublishData ||
+		video.Hidden != permission.Hidden ||
+		video.Recorder != permission.Recorder
+
+	if !hasChanged {
+		return false
+	}
+
+	video.SetCanSubscribe(permission.CanSubscribe)
+	video.SetCanPublish(permission.CanPublish)
+	video.SetCanPublishData(permission.CanPublishData)
+	video.Hidden = permission.Hidden
+	video.Recorder = permission.Recorder
+	p.grants.Store(grants)
+
+	// publish permission has been revoked then remove all published tracks
+	if !video.GetCanPublish() {
```

Comment:
We'll revoke existing tracks when permission is revoked. Note this behavior is different from when subscribe permissions are revoked. It makes a case for changing the behavior for when subscribe permissions are revoked.

IMO it makes sense to unpublish a participant's existing tracks because:
* this is a common scenario for participatory live streaming. When a user is invited to publish and participate, they are given permissions to publish. Afterwards, they are "demoted" back to the audience, where their existing tracks should be unpublished too.
* We don't have other APIs to unpublish from the server.